### PR TITLE
fix(reports): Ensure consistent ASR calculation

### PR DESCRIPTION
**Pull request by will-holley**

### CODE REVIEW

File: src/app/src/pages/redteam/report/components/FrameworkCard.tsx [hunk 1]
```diff
@@ -137,10 +212,10 @@ const FrameworkCard = ({
                   } = categorizePlugins(expandedPlugins, categoryStats, pluginPassRateThreshold);
 
                   // Sort all sets appropriately
-                  const sortedNonCompliantItems = sortedNonCompliantPlugins(
```
Comment (will-holley) [inline src/app/src/pages/redteam/report/components/FrameworkCard.tsx:147]: `sortedNonCompliantPlugins` and `sortedCompliantPlugins` are separate functions with identical logic.

File: src/app/src/utils/redteam.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  }
+}
+
+/**
+ * Formats the ASR for display, applying a consistent number of significant digits and appending a percentage sign.
+ * @param asr - The ASR to format.
+ * @param significantDigits - The number of significant digits to display.
+ * @returns The formatted ASR.
+ */
+export function formatASRForDisplay(asr: number, significantDigits: number = 2): string {
```
Comment (will-holley) [inline src/app/src/utils/redteam.ts:66]: Do we want this to be 2?

File: src/app/src/pages/redteam/report/components/StrategyStats.tsx [hunk 3]
```diff
@@ -208,408 +218,370 @@ const StrategyStats = ({
...
+          <Grid size={{ xs: 4 }}>
+            <Typography variant="h6" align="center" color="error.main">
+              {selectedStrategyStats.pass}
+            </Typography>
+            <Typography variant="body2" align="center" color="text.secondary">
+              Flagged Attempts
+            </Typography>
+          </Grid>
+          <Grid size={{ xs: 4 }}>
+            <Typography variant="h6" align="center">
```
[the remark this answers is not in the document]
Comment (will-holley) [reply src/app/src/pages/redteam/report/components/StrategyStats.tsx:277]: wrong mui version

**Comment by addelong** [comment]
@will-holley conflicts need resolved. Also check for TODOs in comments

File: src/app/src/pages/redteam/report/components/FrameworkCard.tsx [hunk 4]
```diff
@@ -251,63 +275,15 @@ const FrameworkCard = ({
                             </Typography>
                           </ListItem>
                         )}
-                        {sortedNonCompliantItems.map((plugin) => {
```
Comment (addelong) [inline src/app/src/pages/redteam/report/components/FrameworkCard.tsx:254]: i love how much duplication got removed here by extracting the component, nice job

File: src/app/src/utils/redteam.ts (hunk 2 above)
Comment (addelong) [reply src/app/src/utils/redteam.ts:66]: i think any more than that would be basically meaningless to customers, so yes

File: src/app/src/pages/redteam/report/components/StrategyStats.tsx [hunk 5]
```diff
@@ -113,64 +101,86 @@ const StyledStrategyDescription = styled(Typography)(({ theme }) => ({
...
+      });
+    });
+
+    return Object.entries(pluginStats)
+      .map(([plugin, stats]) => ({
+        plugin,
+        ...stats,
+        asr: calculateAttackSuccessRate(stats.total, stats.successfulAttacks),
+      }))
+      .sort((a, b) => b.asr - a.asr);
```
Comment (addelong) [inline src/app/src/pages/redteam/report/components/StrategyStats.tsx]: Very much a nit but we have this same comparator in FrameworkCard.tsx, there might be a way to use the same one in both components.

**Comment by ghost** [comment]
## ❌ Generated 49 tests - 47 passed, 2 failed (69feeb31d1de6f46a8d6437ee609a1dc2a9e09ec) [View tests ↗](https://app.usetusk.ai/app/testing-commit-check/8b7b1f65-d05d-4582-8239-cedc7d847782?client=e91e3433-8fb4-4e98-b233-bba492f4909c)

### Test Summary

- `FrameworkCard` - 6 ✅
- `FrameworkCompliance` - 3 ✅, 1 ❌
- `FrameworkPluginResult` - 6 ✅
- `App` - 7 ✅
- `ReportsDataGrid` - 1 ✅
- `RiskCategories` - 3 ✅, 1 ❌
- `StrategyStats` - 5 ✅
- `formatASRForDisplay` - 8 ✅
- `getASRColor` - 1 ✅
- `getPassRateColor` - 7 ✅

### Results

Tusk's tests are mostly passing with just 2 failures out of 49 tests. The failures are in edge cases related to handling inconsistent data - one in `FrameworkCompliance` where NaN values aren't properly formatted, and another in `RiskCategories` where passed counts exceed total counts. These issues could lead to UI display problems but don't affect the core ASR calculation logic that the PR is fixing.

### Key Points

- ❌ `FrameworkCompliance` fails to handle NaN ASR values properly, displaying "NaN%" instead of "0.00%" when calculation errors occur
- ❌ `RiskCategories` doesn't handle the edge case where `numPassed` exceeds `totalForSubcategory`, potentially causing negative failure counts
- The tests validate that ASR is consistently formatted to 2 decimal places across all components
- `FrameworkPluginResult` tests confirm the component correctly displays plugin results with appropriate icons, colors, and ASR formatting
- Tests verify that ASR calculation correctly uses fail counts (successful attacks) rather than pass counts
- Edge cases like zero tests, very small ASR values, and plugins with identical ASR values are properly handled
- The test suite covers critical paths like report rendering, framework compliance calculation, and risk category aggregation
- Tests confirm that plugins are properly sorted by ASR when data changes, ensuring the most critical issues are highlighted

---

<details><summary>Symbols Not Exported</summary>

The following symbols were skipped because they are not exported and therefore cannot be tested. Export these symbols if you want them to be tested:

- `src/app/src/pages/redteam/report/components/StrategyStats.tsx` - (`DrawerContent`)
- `src/app/src/pages/redteam/report/components/ReportIndex.tsx` - (`ReportsDataGrid`)

</details>
<details><summary>View check history</summary>
<p>

| Commit | Status | Output | Created (UTC) |
|--------|--------|--------|--------|
| aaf4ab76ddd870439442000df1f734b62482d413 | ✅ Generated 15 tests - 15 passed | [Tests](https://app.usetusk.ai/app/testing-commit-check/2833b452-c1a2-432a-a348-0cd6f7e9adb5?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 1, 2025 7:03PM |
| fe68c483c6399256cd20c4d8a5f4fc31419dac3f | ❌ Generated 23 tests - 22 passed, 1 failed | [Tests](https://app.usetusk.ai/app/testing-commit-check/7f99b54a-6dd5-42cc-be62-283b6650fa59?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 1, 2025 7:41PM |
| 6b6cd6bd85abdf394ea25a860db273b47b490bcb | ❌ Generated 23 tests - 22 passed, 1 failed | [Tests](https://app.usetusk.ai/app/testing-commit-check/0451a070-9504-4b65-9122-7448a83ee9d8?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 6, 2025 2:58PM |
| 92157c5e38e1bba17d723b35f7d242c170885575 | ❌ Generated 23 tests - 22 passed, 1 failed | [Tests](https://app.usetusk.ai/app/testing-commit-check/32204d51-61d1-4f37-86e6-2337658debe8?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 6, 2025 3:26PM |
| d89dd8173972967398d750aa140ad19f080f67a2 | ❌ Generated 23 tests - 22 passed, 1 failed | [Tests](https://app.usetusk.ai/app/testing-commit-check/69433db6-ae26-4954-8415-9dbd6f80b692?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 6, 2025 3:43PM |
| c724670d05f18e4eb5a429094aa577990aba36bb | ✅ Generated 1 test - 1 passed | [Tests](https://app.usetusk.ai/app/testing-commit-check/b3cf03ff-5f25-4f62-9729-46e159debd35?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 6, 2025 7:25PM |
| 49e788266fa303d0c9018abf31e84d24001636cf | ✅ Generated 1 test - 1 passed | [Tests](https://app.usetusk.ai/app/testing-commit-check/b6af15e2-bd7b-4a1e-a4a7-ef4032c3172b?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 7, 2025 2:45PM |
| b75f138288bf8c1c3384511f9c6727044aa7c152 | ✅ Generated 1 test - 1 passed | [Tests](https://app.usetusk.ai/app/testing-commit-check/a49fb9d6-add2-4ac4-86f8-c5ee86b9ada6?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 7, 2025 2:49PM |
| fa561005328e323ee2d9af1341cdaa11373fda77 | ✅ Generated 1 test - 1 passed | [Tests](https://app.usetusk.ai/app/testing-commit-check/c3aba47a-7870-49ae-a703-fa8c613fdb88?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 7, 2025 2:52PM |
| 46c01dd72064e7de2bfb1f255f4560db5402178e | ✅ Generated 1 test - 1 passed | [Tests](https://app.usetusk.ai/app/testing-commit-check/0f0c55ef-066c-46b5-9581-63226aa92f5d?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 7, 2025 7:21PM |
| cbeddf5b01bd3d3bcfaecd849058226ac78d1768 | ✅ Generated 1 test - 1 passed | [Tests](https://app.usetusk.ai/app/testing-commit-check/172094ea-5dc8-4bbe-9a27-f17124bfacd2?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 7, 2025 8:16PM |
| e7692e48664fe64b48e265b4ab7f73701d31299c | ⏩ Skipped due to new commit on branch | [Output](https://app.usetusk.ai/app/testing-commit-check/ff4d78c5-c9d7-41a5-a86f-b3de8648def6?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 8, 2025 2:02PM |
| b8c0dcf2f71794e96e57ad595b9d5d7cdc41aec6 | ⏩ Skipped due to new commit on branch | [Output](https://app.usetusk.ai/app/testing-commit-check/bce44413-97d0-4a33-a199-3548b5d89b06?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 8, 2025 2:20PM |
| b0c13b794b0c4510b8e9c287a6ca6794d6f8c1be | ✅ Generated 1 test - 1 passed | [Tests](https://app.usetusk.ai/app/testing-commit-check/01b429d7-85c4-4b80-b17e-b183026e9fff?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 8, 2025 2:52PM |
| 95328699270db2181c72595157a5d1fa14c7d63d | ✅ Generated 1 test - 1 passed | [Tests](https://app.usetusk.ai/app/testing-commit-check/a325420c-bf91-4236-a919-295bf8457d38?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 8, 2025 3:00PM |
| d9be13a404c09d93266a6a18fa99cdc6e077c1bc | ✅ Generated 1 test - 1 passed | [Tests](https://app.usetusk.ai/app/testing-commit-check/9a3a0014-b3f5-4d63-9545-656d2046f016?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 8, 2025 3:12PM |
| 5c5ad36164753c8e2d775399709be63690b4bdb5 | ✅ Generated 1 test - 1 passed | [Tests](https://app.usetusk.ai/app/testing-commit-check/7b2cf589-a516-4ddb-ab6d-fc23a8031423?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 8, 2025 3:29PM |
| 7b00e094ab67ded9ed0e09db6dbb9d494d81b968 | ❌ Generated 45 tests - 43 passed, 2 failed | [Tests](https://app.usetusk.ai/app/testing-commit-check/b6fd8f37-3b9a-4d20-b32d-5f34403f0150?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 8, 2025 3:51PM |
| 47702f17c46660b2ba422edcb376a1d069fe084e | ❌ Generated 45 tests - 43 passed, 2 failed | [Tests](https://app.usetusk.ai/app/testing-commit-check/1a7d184c-6b49-44da-84e7-11e76c0d7f23?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 8, 2025 5:03PM |
| d3feb149e720e01b69dacdeb11e3ba4e08d455ab | ❌ Generated 45 tests - 43 passed, 2 failed | [Tests](https://app.usetusk.ai/app/testing-commit-check/9ff44ca0-6e4f-42ff-b3bf-ebdeaa909ee5?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 8, 2025 5:03PM |
| 95e3f63a2cebb27f46107943731a1aeb39028d55 | ❌ Generated 50 tests - 48 passed, 2 failed | [Tests](https://app.usetusk.ai/app/testing-commit-check/ede921fb-c99c-4c3b-923d-d3a13547d715?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 8, 2025 6:18PM |
| 0603df7d27c09b8dcb187ae80a47e9739c14b094 | ⏩ Test environment setup failed | [Output](https://app.usetusk.ai/app/testing-commit-check/7b243b2b-4509-4d47-b6be-733ac4116315?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 13, 2025 5:49PM |
| 3adb463a76a705f799bec3d024894017d3798da6 | ⏩ Test environment setup failed | [Output](https://app.usetusk.ai/app/testing-commit-check/067376cb-2707-43a6-9664-c39c815e5b91?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 13, 2025 5:53PM |
| db6d01666702e9fa60c3faf57fe4a9307b186a25 | ⏩ Skipped due to new commit on branch | [Output](https://app.usetusk.ai/app/testing-commit-check/a7fcdd10-b2ce-4f66-aedf-7450b6280645?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 13, 2025 9:41PM |
| 3ef3a40586d54c83835f5b782c2f6cc3bb512c3b | ❌ Generated 49 tests - 47 passed, 2 failed | [Tests](https://app.usetusk.ai/app/testing-commit-check/ead0069c-256d-4d79-96ab-1eb6bfef5c2b?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 13, 2025 9:43PM |
| 69feeb31d1de6f46a8d6437ee609a1dc2a9e09ec | ❌ Generated 49 tests - 47 passed, 2 failed | [Tests](https://app.usetusk.ai/app/testing-commit-check/8b7b1f65-d05d-4582-8239-cedc7d847782?client=e91e3433-8fb4-4e98-b233-bba492f4909c) | Oct 13, 2025 10:59PM |

</p>
</details>

[View output in GitHub ↗](https://github.com/promptfoo/promptfoo/pull/5792/checks?check_run_id=52653566468)

_Was Tusk helpful? Give feedback by reacting with 👍 or 👎_

File: src/app/src/pages/redteam/report/components/FrameworkPluginResult.tsx [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  const pluginSeverity =
+    riskCategorySeverityMap[plugin as keyof typeof riskCategorySeverityMap] || Severity.Low;
+
+  return (
+    <ListItem
+      sx={{
+        borderLeft: `3px solid ${getSeverityColor(pluginSeverity, theme)}`,
+        pl: 2,
+        mb: 0.5,
+        bgcolor: 'rgba(0, 0, 0, 0.02)',
```
Comment (addelong) [inline src/app/src/pages/redteam/report/components/FrameworkPluginResult.tsx:48]: does this need to be an `alpha`'d color?

Comment (will-holley) [reply src/app/src/pages/redteam/report/components/FrameworkPluginResult.tsx:48]: Hmmm.... I'm not sure what the correct opaque hex would be since the parent components that define the background color also do so with non-zero alpha.
